### PR TITLE
Add generic KV caching support, use it with Whisper

### PR DESCRIPTION
**Pull request by katalinic-gc**
# What does this PR do?

Intended usage for Whisper:

```python
pipelined_model = pipelined_model.parallelize(
    for_generation=True, use_cache=use_cache, batch_size=batch_size, max_length=448, num_beams=num_beams
)
pipelined_model.generate(input_features, use_cache=use_cache, **kwargs)
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/graphcore/generation_utils.py [hunk 1]
```diff
@@ -63,47 +73,157 @@ def graph_profile_dir_append(append: str):
...
+        self._k_cache.copy_(_k_cache)
+        self._v_cache.copy_(_v_cache)
+
+        # Dynamic update leads to re-arrangement, so use a brute force matmul approach.
+        bsz, heads, src_len, head_dim = self._k_cache.shape
+        mm_mask = (torch.arange(src_len) == self._generation_step).view(1, src_len)
+        _key = torch.matmul(key.view(bsz * heads, head_dim, 1), mm_mask.to(key.dtype))
+        _value = torch.matmul(value.view(bsz * heads, head_dim, 1), mm_mask.to(value.dtype))
+        self._k_cache += _key.permute(0, 2, 1).view(self._k_cache.shape)
+        self._v_cache += _value.permute(0, 2, 1).view(self._v_cache.shape)
```
[the remark this answers is not in the document]
Comment (katalinic-gc) [reply optimum/graphcore/generation_utils.py:139]: I will show you the resulting execution traces.

[the remark this answers is not in the document]
Comment (katalinic-gc) [reply optimum/graphcore/generation_utils.py:137]: We'll go with matmul for empirical reasons for now.

**Comment by katalinic-gc** [comment]
Checking compilation time increase for large

**Review by jimypbr — APPROVED** [review]
Great work!

**Pull request merged**